File: clients/android/sample/src/androidTest/kotlin/dev/arizona/sample/ReconnectE2ETest.kt
package dev.arizona.sample

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import org.junit.Rule
import org.junit.Test

/**
 * On-device (emulator) e2e for reconnect-with-backoff. Navigate to the counter
 * and drive it; force an abrupt socket drop; the client reopens on its own
 * (re-mounting via _az_reconnect=1, so the count resets to 0) and the new socket
 * works -- no relaunch. The Playwright analogue is e2e/native/reconnect.spec.js.
 *
 * Requires the Arizona test server reachable at localhost:4040 (adb reverse
 * tcp:4040 tcp:4040). Run with: ./gradlew :sample:connectedCheck
 */
class ReconnectE2ETest {
    @get:Rule
    val rule = createAndroidComposeRule<MainActivity>()

    @Test
    fun reopensAndRecoversAfterADroppedConnection() {
        // Launcher shows the menu; navigate to the counter.
        rule.waitForText("Counter")
        rule.onNodeWithText("Counter").performClick()
        rule.waitForText("Count: 0")

        // Drive it once so the pre-drop state is distinguishable (count = 1).
        rule.onNodeWithText("+").performClick()
        rule.waitForText("Count: 1")

        // Abruptly drop the socket -> onFailure -> reconnect with backoff. The
        // placeholder flashes during the ~1s gap, then a fresh mount re-renders.
        rule.activity.client.forceDrop()

        // The reconnect re-mounts fresh, so the count resets to 0...
        rule.waitForText("Count: 0")
        // ...and the new socket round-trips.
        rule.onNodeWithText("+").performClick()
        rule.waitForText("Count: 1")
    }
}

File: clients/ios/Sample/UITests/ReconnectE2ETest.swift
import XCTest

/// On-device reconnect-with-backoff -- the analogue of the Android
/// ReconnectE2ETest. Navigate to the counter and drive it; force an abrupt socket
/// drop (the `-uitest` toolbar's "drop" button calls client.forceDrop()); the
/// client reopens on its own (re-mounting via _az_reconnect=1, so the count
/// resets to 0) and the new socket works -- no relaunch.
final class ReconnectE2ETest: XCTestCase {
    private var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = launchSample()
    }

    func testReopensAndRecoversAfterADroppedConnection() {
        tapButton(app, "Counter")
        waitForText(app, "Count: 0")

        // Drive it once so the pre-drop state is distinguishable (count = 1).
        tapButton(app, "+")
        waitForText(app, "Count: 1")

        // Abruptly drop the socket -> reconnect with backoff. The placeholder
        // flashes during the ~1s gap, then a fresh mount re-renders at 0.
        tapButton(app, "dbg-drop")
        waitForText(app, "Count: 0", timeout: 15)

        // The new socket round-trips.
        tapButton(app, "+")
        waitForText(app, "Count: 1")
    }
}
